sources/web/datalab/polymer/components/file-browser/file-browser.ts

```diff
@@ -401,44 +409,60 @@ class FileBrowserElement extends Polymer.Element implements DatalabPageElement {
-    // - Refresh mechanism called by the setInterval().
-    // - User clicking refresh button.
-    // - Files page gaining focus.
-    if (this._fetching) {
-      return this._fileListFetchPromise;
-    }
+  async _fetchFileList(throwOnError = false): Promise<void> {
     if (!this.currentFile) {
-      throw new Error('No current file to retrieve');
+      // No current file to retrieve
+      return Promise.resolve();
     }
+    const fetchFileId = this.currentFile.id;
 
     this._fetching = true;
 
     const hideStatus = this.small || !this._showStatus;
-    this._fileListFetchPromise = this._fileManager.list(this.currentFile.id)
-      .then((newList) => {
-        // Only refresh the UI list if there are any changes. This helps keep
-        // the item list's selections intact most of the time.
-        if (JSON.stringify(this._fileList) !== JSON.stringify(newList)) {
-          this._fileList = newList;
-          this._drawFileList();
-        }
-      })
-      // Now load the sessions and update the running status of each file
-      // whose id is in the session list.
-      // We do not need to load sessions if not displaying the Status column.
-      .then(() => hideStatus ? Promise.resolve([]) : SessionManager.listSessionPaths())
-      .catch((e) => {
-        // Do not block loading files if sessions don't load for some reason.
-        Utils.log.error('Could not load sessions: ' + e.message);
-        return [] as string[];
-      })
-      .then((sessions) => {
+
+    try {
+      const newList = await this._fileManager.list(fetchFileId);
+      // If the current file has changed since this fetch request was made, abort
+      // now. The other request causing the change will eventually update the file
+      // list and unset the _fetching flag.
+      if (fetchFileId !== this.currentFile.id) {
+        return;
+      }
+
+      // Only refresh the UI list if there are any changes. This helps keep
+      // the item list's selections intact most of the time.
+      if (JSON.stringify(this._fileList) !== JSON.stringify(newList)) {
+        this._fileList = newList;
+        this._drawFileList();
+      }
+    } catch (e) {
+      const fileSpec = fetchFileId.toString();
+      const msgPrefix = 'Error getting list of files from ' + fileSpec + ':';
+      if (throwOnError === true) {
+        throw new Error(msgPrefix + ' ' + e.message);
+      } else {
+        Utils.log.error(msgPrefix, e);
+      }
+    }
+
+    // Now load the sessions and update the running status of each file
+    // whose id is in the session list.
+    // We do not need to load sessions when not displaying the Status column,
+    // or if the current fild id has changed since the beginning of this request.
+    if (!hideStatus && fetchFileId === this.currentFile.id) {
+      try {
+        const sessions = await SessionManager.listSessionPaths();
         const listElement = this.$.files as ItemListElement;
```

Comment:
I think you should check for fetchFieldId===this.currentFile.id here, right after getting back the session info, and return if not set, as you do on line 439 just after getting the file list. It looks like it probably won't cause a problem anyway, but I think it would be cleaner to add that bail-out clause.

Reply:
I initially thought this would actually be desirable to have the extra update, since by this point the hard work has already been done (sessions API call). But now I think it's safer to bail out earlier, instead of having multiple calls modifying the polymer properties at the same time.
Done.